File: backends/python/runtime/sce_runtime/scheduler.py
# SPDX-License-Identifier: LGPL-2.1-or-later WITH LicenseRef-SCE-Linking-Exception OR LicenseRef-SCE-Commercial
# SPDX-FileCopyrightText: Copyright (c) 2026 newmassrael
"""W3C SCXML 6.2 delayed-event scheduler with cancel-by-sendid (W3C 6.2.2).

Pull-based: callers (the engine) call `drain_due(now_ms)` to harvest events
whose deadlines have arrived. Virtual time is used so the engine remains
single-threaded and deterministic — wall-clock integration is a caller
concern.
"""

from __future__ import annotations

import heapq
import itertools
from dataclasses import dataclass, field
from typing import Any, Generic, Iterator, List, Optional, Set, TypeVar

E = TypeVar("E")


@dataclass(order=True)
class ScheduledEvent(Generic[E]):
    """One entry in the scheduler's priority queue.

    `due_ms` and `seq` are the ordering keys (stable FIFO on ties);
    `sendid`, `event`, and `data` ride along but do not participate in
    comparison. `data` is the marshalled `<send>` payload (W3C SCXML
    5.10) preserved across the scheduler delay so it surfaces on
    `_event.data` at delivery time.
    """

    due_ms: int
    seq: int
    sendid: str = field(compare=False)
    event: E = field(compare=False)
    data: Any = field(default="", compare=False)
    #: §scxml-6.2.5 — the host-served send this entry performs instead
    #: of delivering `event`, or ``None`` for an ordinary delayed send.
    #:
    #: §scxml-6.2.4 makes a delay a property of the SEND and not of the
    #: processor it named, so a host-served send carrying one is an
    #: ordinary delayed send whose delivery happens to be somebody else's.
    #: Keeping it in THIS queue is what makes that true in practice: one
    #: deadline order across both kinds, one ``<cancel sendid>`` path, one
    #: ``peek_next_due_ms`` answer. A parallel list would oblige every
    #: present and future query to remember it existed.
    host_send: Any = field(default=None, compare=False)


class Scheduler(Generic[E]):
    """Min-heap of `(due_ms, seq)` keyed scheduled events.

    Single-threaded. The caller is responsible for advancing virtual
    time and forwarding drained events into the engine's external queue.
    """

    def __init__(self) -> None:
        self._heap: List[ScheduledEvent[E]] = []
        self._cancelled: Set[str] = set()
        self._counter = itertools.count(1)

    def schedule(
        self,
        due_ms: int,
        sendid: str,
        event: E,
        data: Any = "",
        host_send: Any = None,
    ) -> None:
        """Queue `event` for delivery at `due_ms`. `sendid` identifies the
        entry for later `<cancel>` lookups; empty string ids cannot be
        cancelled (matches W3C SCXML 6.2.2 where `<cancel>` requires a
        sendid). `data` is the marshalled `<send>` payload preserved
        across the delay.

        `host_send` makes the entry a W3C SCXML 6.2.5 host-served send to
        be PERFORMED at `due_ms` rather than an event to be delivered —
        the same queue, so it is ordered against every other delayed send
        by deadline and cancelled by the same id."""
        heapq.heappush(
            self._heap,
            ScheduledEvent(
                due_ms=due_ms,
                seq=next(self._counter),
                sendid=sendid,
                event=event,
                data=data,
                host_send=host_send,
            ),
        )

    def cancel(self, sendid: str) -> None:
        """W3C SCXML 6.2.2 — mark `sendid` cancelled; the matching entry
        is skipped the next time it would be drained. No-op on empty
        `sendid` (matches the W3C "id must be set to cancel" semantics)."""
        if sendid:
            self._cancelled.add(sendid)

    def drain_due(self, now_ms: int) -> Iterator[ScheduledEvent[E]]:
        """Yield every scheduled event whose `due_ms <= now_ms`, popping
        them off the heap. Cancelled entries are silently discarded as
        they would have been delivered.

        Draining is a generator, so a caller that runs a macrostep per
        entry sees each cancellation the previous one performed — see
        `pop_due`, which is the one-at-a-time form the engine uses."""
        while self._heap and self._heap[0].due_ms <= now_ms:
            entry = heapq.heappop(self._heap)
            if entry.sendid and entry.sendid in self._cancelled:
                self._cancelled.discard(entry.sendid)
                continue
            yield entry

    def pop_due(self, now_ms: int) -> Optional[ScheduledEvent[E]]:
        """Pop the single earliest entry due at `now_ms`, or None.

        The engine dispatches one of these per macrostep so a `<cancel>`
        performed by an earlier event still reaches a later one that has
        not been delivered yet. Popping them all at once instead makes
        every later entry undroppable, which is how a settle timer —
        arm a long `<send delay>`, cancel it when the short signal
        arrives first — delivers the event it was told to cancel."""
        for entry in self.drain_due(now_ms):
            return entry
        return None

    def peek_next_due_ms(self) -> Optional[int]:
        """The `due_ms` of the earliest entry that would actually be
        delivered, or None if there is none.

        Cancelled entries stay on the heap until a drain walks past
        them, so the front of the heap is not necessarily a deadline
        anyone will see; they are dropped here first. Callers use this
        to compute the next wake deadline — see
        `Engine.time_until_next_scheduled_ms`."""
        while self._heap:
            head = self._heap[0]
            if head.sendid and head.sendid in self._cancelled:
                heapq.heappop(self._heap)
                self._cancelled.discard(head.sendid)
                continue
            return head.due_ms
        return None

    def __len__(self) -> int:
        return len(self._heap)

File: backends/go/runtime/scheduler.go
// SPDX-License-Identifier: LGPL-2.1-or-later WITH LicenseRef-SCE-Linking-Exception OR LicenseRef-SCE-Commercial
// SPDX-FileCopyrightText: Copyright (c) 2025 newmassrael

package sce

import "fmt"

// PullScheduler manages delayed event scheduling (§scxml-6.2).
//
// 1:1 port of Rust PullScheduler<E> from backends/rust/runtime/src/engine.rs.
// Stores delayed events with their ready-at time. The engine polls via
// PopReadyEvent during tick() to fire events whose delay has elapsed.
type PullScheduler[E any] struct {
	entries        []scheduledEntry[E]
	nextAutoSendID uint64
}

// scheduledEntry is a single scheduled event with its metadata.
//
// readyAtMs is milliseconds on the engine's SceClock, not a time.Time: the
// queue must not know which clock the engine reads, or it would be reading one
// of its own and the host could not own time. Every method below is handed the
// reading instead of taking one.
type scheduledEntry[E any] struct {
	event     E
	eventData string
	sendID    string
	readyAtMs int64
}

// NewPullScheduler constructs an empty scheduler.
func NewPullScheduler[E any]() *PullScheduler[E] {
	return &PullScheduler[E]{
		entries: make([]scheduledEntry[E], 0, 4),
	}
}

// ScheduleEventAt queues an event to come due at readyAtMs on the engine's
// clock (§scxml-6.2).
//
// If sendID is empty, an automatic ID is generated. Returns the ID used
// (caller can use it to cancel). Matches Rust PullScheduler::schedule_event_at:
// the deadline is resolved by the caller, which is the only party that knows
// which clock the engine reads.
func (s *PullScheduler[E]) ScheduleEventAt(event E, readyAtMs int64, sendID, eventData string) string {
	effectiveSendID := sendID
	if effectiveSendID == "" {
		s.nextAutoSendID++
		effectiveSendID = fmt.Sprintf("auto_send_%d", s.nextAutoSendID)
	}
	s.entries = append(s.entries, scheduledEntry[E]{
		event:     event,
		eventData: eventData,
		sendID:    effectiveSendID,
		readyAtMs: readyAtMs,
	})
	return effectiveSendID
}

// CancelEvent cancels a scheduled event by send ID (§scxml-6.2.5).
// Returns true if the event was found and removed.
//
// Matches Rust PullScheduler::cancel_event.
func (s *PullScheduler[E]) CancelEvent(sendID string) bool {
	before := len(s.entries)
	n := 0
	for _, e := range s.entries {
		if e.sendID != sendID {
			s.entries[n] = e
			n++
		}
	}
	s.entries = s.entries[:n]
	return n < before
}

// HasReadyEventsAt returns whether any scheduled events are ready to fire
// (readyAtMs <= nowMs).
//
// Matches Rust PullScheduler::has_ready_events_at.
func (s *PullScheduler[E]) HasReadyEventsAt(nowMs int64) bool {
	for _, e := range s.entries {
		if e.readyAtMs <= nowMs {
			return true
		}
	}
	return false
}

// PopReadyEventAt pops the ready event that came due first and its data, judged
// against nowMs. Returns (event, data, true) if an event is ready, or
// (zero, "", false) if nothing is.
//
// Deadline order, not insertion order: the caller dispatches these one at a
// time and runs a macrostep between them, so whichever comes out first is the
// one whose transitions run first. Picking by insertion would let a
// later-scheduled event be delivered ahead of an earlier one whenever the host
// woke after both came due, which is the difference between a <cancel> landing
// and being lost. Ties keep insertion order.
//
// Matches Rust PullScheduler::pop_ready_event_at.
func (s *PullScheduler[E]) PopReadyEventAt(nowMs int64) (E, string, bool) {
	best := -1
	for i, e := range s.entries {
		if e.readyAtMs > nowMs {
			continue
		}
		if best < 0 || e.readyAtMs < s.entries[best].readyAtMs {
			best = i
		}
	}
	if best < 0 {
		var zero E
		return zero, "", false
	}
	entry := s.entries[best]
	s.entries = append(s.entries[:best], s.entries[best+1:]...)
	return entry.event, entry.eventData, true
}

// HasPendingEvents returns whether there are any scheduled events (ready or not).
func (s *PullScheduler[E]) HasPendingEvents() bool {
	return len(s.entries) > 0
}

// NextReadyAtMs reports when the earliest still-queued entry comes due, in
// milliseconds on the engine's clock, whether or not it is ready yet. The bool
// is false when nothing is scheduled.
//
// The queue has always known this; nothing could ask. A host driving the
// machine has to decide when to call Tick again, and without this it can only
// guess an interval — see Engine.TimeUntilNextScheduled for what guessing
// costs. Matches Rust PullScheduler::next_ready_at.
func (s *PullScheduler[E]) NextReadyAtMs() (int64, bool) {
	if len(s.entries) == 0 {
		return 0, false
	}
	next := s.entries[0].readyAtMs
	for _, e := range s.entries[1:] {
		if e.readyAtMs < next {
			next = e.readyAtMs
		}
	}
	return next, true
}
